### PR TITLE
Revert the test_mixins to test image by image

### LINKED ISSUE #5211: Error while train default Faster R-CNN model
RuntimeError: operation does not have an identity.
It can work sometimes
2021-05-24 15-06-54屏幕截图

### PR DESCRIPTION
**Pull request by jshilong**
## Motivation
In order to support batch inference in ONNX,  PR#3853 adding dummy proposal and batch dimension make the code difficult to
understand. 
## Modification
This PR modifies the `test_mixins.py` and `bbox_head.py` to remove all batch-inference-related code and be consistent with the logic before PR#3853.

Speed Comparison with 1080TI
Model\FPS | batch parallel(batch1/batch4) FPS | image by image(batch1/batch4) FPS
-- | -- | --
Faster-RCNN batch_size | 10.1/8.5 | 10.2/8.5
Mask-RCNN | 7.5/6.7 | 7.7/6.7
Cascade-RCNN | 7.6/6.6 | 7.5/6.6
Cacade-Mask-RCNN | 5.3/4.9 | 5.3/4.9

## NOTES
When input proposals of `batch_nms` has shape (0, 4), `rpn_head` would raise an error
```
    max_coordinate = boxes.max()
RuntimeError: operation does not have an identity.
```
This PR fix this.
Closes #5211

### CODE REVIEW

File: mmdet/models/roi_heads/test_mixins.py [hunk 1]
```diff
@@ -207,7 +166,9 @@ async def async_test_mask(self,
             if det_bboxes.shape[0] == 0:
                 segm_result = [[] for _ in range(self.mask_head.num_classes)]
             else:
-                if rescale:
+                # TODO: check this @haian
+                if rescale and not isinstance(scale_factor,
+                                              (float, torch.Tensor)):
```
Comment (hhaAndroid) [inline mmdet/models/roi_heads/test_mixins.py:171]: remove float

Comment (jshilong) [reply mmdet/models/roi_heads/test_mixins.py:171]: Would it be float?

Comment (jshilong) [reply mmdet/models/roi_heads/test_mixins.py:170]: Would it be a Tensor?

File: mmdet/models/dense_heads/rpn_head.py [hunk 2]
```diff
@@ -75,39 +76,105 @@ def loss(self,
...
+            with_nms (bool): If True, do nms before return boxes.
+                Default: True.
 
         Returns:
             list[tuple[Tensor, Tensor]]: Each item in result_list is 2-tuple.
                 The first item is an (n, 5) tensor, where the first 4 columns
                 are bounding box positions (tl_x, tl_y, br_x, br_y) and the
                 5-th column is a score between 0 and 1. The second item is a
-                (n,) tensor where each item is the predicted class label of the
+                (n,) tensor where each item is the predicted class labelof the
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rpn_head.py:108]: labelof ？

File: mmdet/models/roi_heads/test_mixins.py (hunk 1 above)
Comment (hhaAndroid) [reply mmdet/models/roi_heads/test_mixins.py:170]: It is best to be able to comment on which algorithms use float and which use tensor.

File: mmdet/models/dense_heads/rpn_head.py [hunk 3]
```diff
@@ -75,39 +76,105 @@ def loss(self,
...
+            img_shape = img_metas[img_id]['img_shape']
+            scale_factor = img_metas[img_id]['scale_factor']
+            if with_nms:
+                # some heads don't support with_nms argument
+                proposals = self._get_bboxes_single(cls_score_list,
+                                                    bbox_pred_list,
+                                                    mlvl_anchors, img_shape,
+                                                    scale_factor, cfg, rescale)
+            else:
+                proposals = self._get_bboxes_single(cls_score_list,
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rpn_head.py:137]: This function does not need.

File: mmdet/models/dense_heads/rpn_head.py [hunk 4]
```diff
@@ -75,39 +76,105 @@ def loss(self,
...
+                           bbox_preds,
+                           mlvl_anchors,
+                           img_shape,
+                           scale_factor,
+                           cfg,
+                           rescale=False):
+        """Transform outputs for a single batch item into bbox predictions.
+
+          Args:
+            cls_score_list (list[Tensor]): Box scores of all scale level
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rpn_head.py:156]: docstr error

File: mmdet/models/dense_heads/rpn_head.py [hunk 5]
```diff
@@ -75,39 +76,105 @@ def loss(self,
...
+                each item has shape (num_total_anchors, 4).
+            img_shape (tuple[int]): Shape of the input image,
+                (height, width, 3).
+            scale_factor (ndarray): Scale factor of the image arrange as
+                (w_scale, h_scale, w_scale, h_scale).
+            cfg (mmcv.Config): Test / postprocessing configuration,
+                if None, test_cfg would be used.
+            rescale (bool): If True, return boxes in original image space.
+                Default: False.
+            with_nms (bool): If True, do nms before return boxes.
```
Comment (hhaAndroid) [inline mmdet/models/dense_heads/rpn_head.py:170]: with_nms

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 6]
```diff
@@ -317,122 +316,58 @@ def get_bboxes(self,
...
-        elif rois.ndim == 3:
-            # all input tensor have batch dimension
-            batch_mode = True
-            assert isinstance(scale_factor, tuple)
-        else:
-            raise NotImplementedError(f'Unexpect shape of roi {rois.shape}')
-
-        # bbox_pred would be None in some detector when with_reg is False,
-        # e.g. Grid R-CNN.
         if bbox_pred is not None:
```
Comment (hhaAndroid) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:349]: The comment should not be deleted

**Comment by ZwwWayne** [comment]
This PR involves unrelated modifications, please clean them before code review.

File: mmdet/models/dense_heads/rpn_head.py [hunk 7]
```diff
@@ -74,30 +75,29 @@ def loss(self,
...
-                    img_shapes,
-                    scale_factors,
-                    cfg,
-                    rescale=False):
-        """Transform outputs for a single batch item into bbox predictions.
+    @force_fp32(apply_to=('cls_scores', 'bbox_preds'))
+    def get_bboxes(self,
+                   cls_scores,
+                   bbox_preds,
+                   img_metas,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:82]: mlvl_anchors is gone, which introduces BC-breaking. Need to handle that.

File: mmdet/models/roi_heads/bbox_heads/bbox_head.py [hunk 8]
```diff
@@ -317,122 +316,59 @@ def get_bboxes(self,
...
-            assert isinstance(scale_factor, tuple)
-        else:
-            raise NotImplementedError(f'Unexpect shape of roi {rois.shape}')
-
         # bbox_pred would be None in some detector when with_reg is False,
         # e.g. Grid R-CNN.
         if bbox_pred is not None:
             bboxes = self.bbox_coder.decode(
-                rois[..., 1:], bbox_pred, max_shape=img_shape)
+                rois[:, 1:], bbox_pred, max_shape=img_shape)
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/bbox_heads/bbox_head.py:352]: Can we keep `rois[...,]` as this is more general?

Comment (jshilong) [reply mmdet/models/roi_heads/bbox_heads/bbox_head.py:352]: sure

**Comment by jshilong** [comment]
I need to test all Two-Stage models to make sure all the model is as usual.

**Comment by jshilong** [comment]
> I need to test all Two-Stage models to make sure all the model is as usual.

Done

File: mmdet/models/dense_heads/rpn_head.py [hunk 9]
```diff
@@ -115,77 +170,59 @@ def _get_bboxes(self,
...
-            result_list.append(dets[:cfg.max_per_img])
-        return result_list
+        if cfg.min_bbox_size > 0:
+            w = proposals[:, 2] - proposals[:, 0]
+            h = proposals[:, 3] - proposals[:, 1]
+            valid_inds = torch.nonzero(
+                (w >= cfg.min_bbox_size)
+                & (h >= cfg.min_bbox_size),
+                as_tuple=False).squeeze()
+            if valid_inds.sum().item() != len(proposals):
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:218]: `valid_inds.sum().item() != len(proposals)` -> `not valid_inds.all()`

File: mmdet/models/dense_heads/rpn_head.py [hunk 10]
```diff
@@ -115,77 +170,59 @@ def _get_bboxes(self,
...
+                proposals = proposals[valid_inds, :]
+                scores = scores[valid_inds]
+                ids = ids[valid_inds]
+
+        dets, keep = batched_nms(proposals, scores, ids, cfg.nms)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:223]: batched_nms support to set `max_num` now. So we can considering update cfg.nms without modify dets manually. see [here](https://github.com/open-mmlab/mmcv/blob/master/mmcv/ops/nms.py#L316).

File: mmdet/models/dense_heads/rpn_head.py (hunk 9 above)
Comment (jshilong) [reply mmdet/models/dense_heads/rpn_head.py:218]: `valid_inds` is not a binary mask.

Comment (jshilong) [reply mmdet/models/dense_heads/rpn_head.py:218]: I would change `valid_inds` to a binary mask.

File: mmdet/models/dense_heads/rpn_head.py (hunk 10 above)
Comment (jshilong) [reply mmdet/models/dense_heads/rpn_head.py:223]: Seems I want to deprecated `max_num` in #4636

Comment (jshilong) [reply mmdet/models/dense_heads/rpn_head.py:223]: https://github.com/open-mmlab/mmdetection/blob/3ea7905418da1c26a600ef26016ee66d21cc733f/mmdet/core/post_processing/bbox_nms.py#L87
We haven't used  `max_num` in `batched_nms` before.

File: mmdet/models/dense_heads/rpn_head.py [hunk 11]
```diff
@@ -115,77 +170,59 @@ def _get_bboxes(self,
...
+            h = proposals[:, 3] - proposals[:, 1]
+            valid_mask = (w >= cfg.min_bbox_size) & (h >= cfg.min_bbox_size)
+            if not valid_mask.all():
+                proposals = proposals[valid_mask]
+                scores = scores[valid_mask]
+                ids = ids[valid_mask]
+        if len(proposals):
+            dets, keep = batched_nms(proposals, scores, ids, cfg.nms)
+        else:
+            return proposals.new_zeros(0, 5)
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:222]: Can we directly return proposals?

File: mmdet/models/dense_heads/rpn_head.py [hunk 12]
```diff
@@ -115,77 +170,59 @@ def _get_bboxes(self,
...
-        return result_list
+        if cfg.min_bbox_size > 0:
+            w = proposals[:, 2] - proposals[:, 0]
+            h = proposals[:, 3] - proposals[:, 1]
+            valid_mask = (w >= cfg.min_bbox_size) & (h >= cfg.min_bbox_size)
+            if not valid_mask.all():
+                proposals = proposals[valid_mask]
+                scores = scores[valid_mask]
+                ids = ids[valid_mask]
+        if len(proposals):
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:219]: proposals.numel() > 0

File: mmdet/models/roi_heads/test_mixins.py [hunk 13]
```diff
@@ -244,69 +204,50 @@ def simple_test_mask(self,
         ori_shapes = tuple(meta['ori_shape'] for meta in img_metas)
         scale_factors = tuple(meta['scale_factor'] for meta in img_metas)
 
+        if isinstance(scale_factors[0], float):
+            logger.warning(
+                'Scale factor in img_metas should be a '
+                'ndarray with shape (4,) '
+                'arrange as (factor_w, factor_h, factor_w, factor_h), '
+                'The scale_factor with float type has been deprecated. ')
+            scale_factors = np.array([scale_factors] * 4)
```
Comment (ZwwWayne) [inline mmdet/models/roi_heads/test_mixins.py:213]: why not directly create tensor?

File: mmdet/models/dense_heads/rpn_head.py (hunk 11 above)
Comment (jshilong) [reply mmdet/models/dense_heads/rpn_head.py:222]: proposals has shape (0, 4)

**Review by ZwwWayne — APPROVED** [review]

**Pull request merged**